src/types/poly_func.rs

```diff
@@ -0,0 +1,524 @@
+                        Type::new_var_use(0, TypeBound::Copyable),
+                        // Next is the free variable that we substituted in (hence Eq)
+                        // - renumbered because of the intervening forall (Copyable)
+                        Type::new_var_use(FREE + 1, TypeBound::Eq)
+                    )
+                ))]
+            )
+        );
+
+        // Also try substituting in a type containing both free and bound vars
+        let rhs = |i| {
+            Type::new_function(new_pf1(
+                TP_EQ,
+                Type::new_var_use(0, TypeBound::Eq),
+                new_array(
+                    Type::new_var_use(0, TypeBound::Eq),
+                    TypeArg::new_var_use(i, TypeParam::max_nat()),
+                ),
+            ))
+        };
+
+        let res = pf
+            .instantiate_all(&[TypeArg::Type { ty: rhs(FREE) }], &reg)
+            .unwrap();
+        assert_eq!(
+            res,
+            FunctionType::new(
+                vec![rhs(FREE)],
+                vec![Type::new_function(new_pf1(
+                    TypeParam::Type(TypeBound::Copyable),
+                    Type::new_var_use(0, TypeBound::Copyable),
+                    list_of_tup(
+                        Type::new_var_use(0, TypeBound::Copyable), // not renumbered...
+                        rhs(FREE + 1)                              // renumbered
+                    )
+                ))]
+            )
+        )
+    }
+
+    const USIZE_TA: TypeArg = TypeArg::Type { ty: USIZE_T };
+
+    #[test]
+    fn test_instantiate() -> Result<(), SignatureError> {
+        let array_max = PolyFuncType::new_validated(
+            vec![TypeParam::Type(TypeBound::Any), TypeParam::max_nat()],
+            FunctionType::new(
+                vec![new_array(
+                    Type::new_var_use(0, TypeBound::Any),
+                    TypeArg::new_var_use(1, TypeParam::max_nat()),
+                )],
+                vec![Type::new_var_use(0, TypeBound::Any)],
+            ),
+            &PRELUDE_REGISTRY,
+        )?;
+
+        let concrete = FunctionType::new(
+            vec![new_array(USIZE_T, TypeArg::BoundedNat { n: 3 })],
+            vec![USIZE_T],
+        );
+        let actual =
+            array_max.instantiate(&[USIZE_TA, TypeArg::BoundedNat { n: 3 }], &PRELUDE_REGISTRY)?;
+
+        assert_eq!(actual, concrete.into());
+
+        let partial = PolyFuncType::new_validated(
+            vec![TypeParam::max_nat()],
+            FunctionType::new(
+                vec![new_array(
+                    USIZE_T,
+                    TypeArg::new_var_use(0, TypeParam::max_nat()),
+                )],
+                vec![USIZE_T],
+            ),
+            &PRELUDE_REGISTRY,
+        )?;
+        let res = array_max.instantiate(&[USIZE_TA], &PRELUDE_REGISTRY)?;
+        assert_eq!(res, partial);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_type_apply_nested() -> Result<(), SignatureError> {
```

Comment:
What's the difference between this test and the first case in `substitute_under_binder`?

Reply:
So it's `forall A. A -> (forall C. C -> List(Tuple(C, A))` vs `forall N. -> (forall A. Array(A, N) -> A)`. Isn't this mostly the same?

Reply:
Ooof! Yes, you are right. (The later tests were originally in `type_apply.rs` and then as I shifted more logic from there to here I realized that the tests should move too....well spotted, sir!)

Ok, so I have commoned up into a helper function `nested_func()` used in both. Saves quite a few lines, but the two tests remain (one is testing more complex RHSes than the other). Thanks! :)